File: src/main/java/net/badgersmc/em/EnthusiaMarketPluginLoader.java
package net.badgersmc.em;

import io.papermc.paper.plugin.loader.PluginClasspathBuilder;
import io.papermc.paper.plugin.loader.PluginLoader;
import io.papermc.paper.plugin.loader.library.impl.MavenLibraryResolver;
import org.eclipse.aether.artifact.DefaultArtifact;
import org.eclipse.aether.graph.Dependency;
import org.eclipse.aether.repository.RemoteRepository;
import org.jetbrains.annotations.NotNull;

/**
 * Declares runtime libraries against repo1.maven.org directly so we don't
 * depend on whatever Maven Central mirror the server happens to be configured
 * with (Leaf forks hardcode maven.aliyun.com, which has had outages).
 *
 * IMPORTANT: this class must be Java, not Kotlin. PluginLoader.classloader()
 * runs before any runtime libraries are resolved — it's the thing that
 * schedules them — so the loader's bytecode cannot reference kotlin-stdlib
 * classes (Intrinsics, etc). Paper loads this in an isolated early classloader
 * that only sees paper-api plus the plugin jar's own classes.
 *
 * Keep this list in sync with the shadowJar excludes in build.gradle.kts. If
 * you add a runtime dep, exclude it there too — otherwise it ships in the fat
 * jar AND gets downloaded at runtime.
 */
@SuppressWarnings("UnstableApiUsage")
public class EnthusiaMarketPluginLoader implements PluginLoader {

    @Override
    public void classloader(@NotNull PluginClasspathBuilder builder) {
        MavenLibraryResolver resolver = new MavenLibraryResolver();

        resolver.addRepository(
            new RemoteRepository.Builder(
                "central",
                "default",
                "https://repo1.maven.org/maven2/"
            ).build()
        );

        String[] coords = {
            "com.zaxxer:HikariCP:5.1.0",
            "org.xerial:sqlite-jdbc:3.45.1.0",
            "org.mariadb.jdbc:mariadb-java-client:3.3.2",
            "org.slf4j:slf4j-nop:2.0.13",
            "org.jetbrains.kotlin:kotlin-stdlib:2.0.21",
            "org.jetbrains.kotlin:kotlin-reflect:2.0.21",
            "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.0",
            "com.charleskorn.kaml:kaml:0.59.0"
        };

        for (String coord : coords) {
            resolver.addDependency(new Dependency(new DefaultArtifact(coord), null));
        }

        builder.addLibrary(resolver);
    }
}

File: src/main/kotlin/net/badgersmc/em/EnthusiaMarketPluginLoader.kt
package net.badgersmc.em

import io.papermc.paper.plugin.loader.PluginClasspathBuilder
import io.papermc.paper.plugin.loader.PluginLoader
import io.papermc.paper.plugin.loader.library.impl.MavenLibraryResolver
import org.eclipse.aether.artifact.DefaultArtifact
import org.eclipse.aether.graph.Dependency
import org.eclipse.aether.repository.RemoteRepository

/**
 * Declares runtime libraries explicitly against `repo1.maven.org` instead of
 * relying on the server's configured Maven Central URL.
 *
 * Why this exists: plugin.yml's `libraries:` block resolves through whatever the
 * server admin (or fork author) set as the central repository. Leaf — and some
 * other Paper forks — hardcode `maven.aliyun.com` as central, and when that
 * mirror has an outage every plugin using runtime libraries fails to load with
 * a 502 / connection error. A `PluginLoader` runs before classloading and lets
 * us specify the repository URL ourselves, sidestepping the server config.
 *
 * Keep this list in sync with paper-plugin.yml (the file is purely informational —
 * Paper reads its libraries from this loader, not from the yml). If you add or
 * remove a runtime dep, update build.gradle.kts's shadowJar excludes too.
 */
@Suppress("UnstableApiUsage")
class EnthusiaMarketPluginLoader : PluginLoader {

    override fun classloader(builder: PluginClasspathBuilder) {
        val resolver = MavenLibraryResolver()

        resolver.addRepository(
            RemoteRepository.Builder(
                "central",
                "default",
                "https://repo1.maven.org/maven2/"
            ).build()
        )

        listOf(
            "com.zaxxer:HikariCP:5.1.0",
            "org.xerial:sqlite-jdbc:3.45.1.0",
            "org.mariadb.jdbc:mariadb-java-client:3.3.2",
            "org.slf4j:slf4j-nop:2.0.13",
            "org.jetbrains.kotlin:kotlin-stdlib:2.0.21",
            "org.jetbrains.kotlin:kotlin-reflect:2.0.21",
            "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.0",
            "com.charleskorn.kaml:kaml:0.59.0"
        ).forEach { coord ->
            resolver.addDependency(Dependency(DefaultArtifact(coord), null))
        }

        builder.addLibrary(resolver)
    }
}
